brave/src/test/java/com/linecorp/armeria/server/brave/BraveServiceIntegrationTest.java

```diff
@@ -65,23 +78,63 @@ protected void init() {
-        sb.service("/extra",
+        sb.service("/async", (ctx, req) -> asyncResponse(future ->
+            future.complete(HttpResponse.of(OK, MediaType.PLAIN_TEXT_UTF_8, "bar"))));
+
+        sb.service("/exception", (ctx, req) -> {
+            // TODO: how do we set status 503 and also retain the cause's message?
+            throw new IllegalStateException("not ready");
+        });
+        sb.service("/exceptionAsync", (ctx, req) -> asyncResponse(future ->
+            // TODO: how do we set status 503 and also retain the cause's message?
+            future.completeExceptionally(new IllegalStateException("not ready"))));
+
+        sb.service("/items/:itemId",
                    (ctx, req) -> HttpResponse.of(OK, MediaType.PLAIN_TEXT_UTF_8,
-                                                 String.valueOf(req.headers().get(EXTRA_KEY))));
-        sb.service("/badrequest", (ctx, req) -> HttpResponse.of(BAD_REQUEST));
+                                                 String.valueOf(ctx.pathParam("itemId"))));
+        sb.service("/async_items/:itemId", (ctx, req) -> asyncResponse(future ->
+            future.complete(HttpResponse.of(OK, MediaType.PLAIN_TEXT_UTF_8,
+                                                String.valueOf(ctx.pathParam("itemId"))))));
+        // TODO: how do we mount "/items/:itemId" under the prefix "/nested"?
+
         sb.service("/child", (ctx, req) -> {
+            // TODO: this fails because the timestamp is out of range, eventhough it finishes before
+            // the response is returned.
             Tracing.currentTracer().nextSpan().name("child").start().finish();
             return HttpResponse.of(OK, MediaType.PLAIN_TEXT_UTF_8, "happy");
         });
-        sb.service("/exception", (ctx, req) -> {
-            throw new Exception("not ready");
+        sb.service("/extra",
+            (ctx, req) -> HttpResponse.of(OK, MediaType.PLAIN_TEXT_UTF_8,
+                String.valueOf(req.headers().get(EXTRA_KEY))));
+        sb.service("/badrequest", (ctx, req) -> HttpResponse.of(BAD_REQUEST));
+        sb.service("/child", (ctx, req) -> {
+            Tracing.currentTracer().nextSpan().name("child").start().finish();
+            return HttpResponse.of(OK, MediaType.PLAIN_TEXT_UTF_8, "happy");
         });
+
         sb.decorator(BraveService.newDecorator(httpTracing));
 
         server = sb.build();
         server.start().join();
     }
 
+    HttpResponse asyncResponse(Consumer<CompletableFuture<HttpResponse>> completeResponse) {
+        final CompletableFuture<HttpResponse> responseFuture = new CompletableFuture<>();
+        final HttpResponse res = HttpResponse.from(responseFuture);
+        executorService.submit(() -> completeResponse.accept(responseFuture));
+        return res;
+    }
+
+    @Override
+    @Test
+    public void createsChildSpan() {
+        // Armeria uses different timings than Tracing.clock(context) provided by Brave. This means
```

Comment:
chatted about this with @anuraaga.. maybe we can share the clock used for the log timestamps in armeria with brave perhaps..